### PR TITLE
Upgrade to liberty

### PR DESCRIPTION
**Pull request by ovisan**
Upgrade nova and neutron to liberty

### CODE REVIEW

File: chef/cookbooks/hyperv/recipes/openstack_install_neutron.rb [hunk 1]
```diff
@@ -10,7 +10,6 @@
   #{node[:sevenzip][:command]} x #{node[:cache_location]}#{node[:openstack][:neutron][:file]} -o#{node[:openstack][:location]} -r -y
   #{node[:sevenzip][:command]} x #{node[:openstack][:location]}\\dist\\#{node[:openstack][:neutron][:name]}-#{node[:openstack][:neutron][:version]}.tar -o#{node[:openstack][:location]} -r -y
   rmdir /S /Q #{node[:openstack][:location]}\\dist
-  ren #{node[:openstack][:location]}\\#{node[:openstack][:neutron][:name]}-#{node[:openstack][:neutron][:version]} #{node[:openstack][:neutron][:name]}
```
Comment (vuntz) [inline chef/cookbooks/hyperv/recipes/openstack_install_neutron.rb:13]: Why are you removing this? The `not_if` line will have no effect anymore.

File: chef/cookbooks/hyperv/recipes/openstack_install_neutron.rb [hunk 2]
```diff
@@ -25,3 +24,13 @@
   not_if { ::File.exists?("#{node[:openstack][:neutron][:installed]}") }
 end
 
+ruby_block "ensure correct python path in shebang" do
```
Comment (vuntz) [inline chef/cookbooks/hyperv/recipes/openstack_install_neutron.rb]: What do you have in the shebang?

Comment (ovisan) [inline chef/cookbooks/hyperv/recipes/openstack_install_neutron.rb]: `#!C:\Python27_HyperVNovaCompute\python.exe`
No idea why this path was used, but having the shebang replaced with our python path can only avoid this in the future.

**Comment by tboerger** [comment]
@vuntz what about ceilometer update?

**Comment by tboerger** [comment]
LGTM codewise

**Comment by vuntz** [comment]
> @vuntz what about ceilometer update?

that's for @ovisan, not me, I guess

**Comment by tboerger** [comment]
> that's for @ovisan, not me, I guess

@ovisan said it's not part of the card.

**Comment by vuntz** [comment]
> > that's for @ovisan, not me, I guess
> 
> @ovisan said it's not part of the card.

I would assume it should be. Something to discuss with @jdsn...

**Comment by tboerger** [comment]
So @jdsn will create a followup card for ceilometer. Than I'm fine with this PR as ceilometer will be done separately.

File: chef/cookbooks/hyperv/recipes/openstack_install_nova.rb [hunk 3]
```diff
@@ -24,3 +24,14 @@
   EOH
   not_if { ::File.exists?("#{node[:openstack][:nova][:installed]}") }
 end
+
+ruby_block "ensure correct python path in shebang" do
```
Comment (jdsn) [inline chef/cookbooks/hyperv/recipes/openstack_install_nova.rb]: this looks like code duplication
can this be moved to a helper and be used from both places?

Comment (tboerger) [inline chef/cookbooks/hyperv/recipes/openstack_install_nova.rb]: The only thing I can think about should be another resource at https://github.com/crowbar/crowbar-core/tree/master/chef/cookbooks/utils/providers and https://github.com/crowbar/crowbar-core/tree/master/chef/cookbooks/utils/resources

Comment (ovisan) [inline chef/cookbooks/hyperv/recipes/openstack_install_nova.rb]: This implies more refactoring and dependency to https://github.com/crowbar/crowbar-core
The function would be to search and replace a line in a file, the search being a regexp and the replacement as string. What we need to do is to discuss if it's worth to have this function for future use and estimate the effort in a new card.

Comment (jdsn) [inline chef/cookbooks/hyperv/recipes/openstack_install_nova.rb]: ok, then keep it as it is but please add two comments in the code that points to the both places of this code, so that it gets obvious where we need to refactor in the future

Comment (ovisan) [inline chef/cookbooks/hyperv/recipes/openstack_install_nova.rb]: @jdsn it could be useful to have a function to search and replace a line added to core, so if a card will be created, i can work on it. It would have an effort of 2 maximum.

File: chef/cookbooks/hyperv/recipes/openstack_install_neutron.rb [hunk 4]
```diff
@@ -25,3 +25,9 @@
   not_if { ::File.exists?("#{node[:openstack][:neutron][:installed]}") }
 end
 
+utils_line "ensure correct python path in shebang" do
+  file "#{node[:python][:scripts]}\\#{node[:service][:neutron][:name]}-script.py"
+  regexp /\A#!.*/
+  replace "#! #{node[:python][:command]}"
+  action :replace
```
Comment (ovisan) [inline chef/cookbooks/hyperv/recipes/openstack_install_neutron.rb:32]: depends on https://github.com/crowbar/crowbar-core/pull/96

File: chef/cookbooks/hyperv/recipes/openstack_install_nova.rb (hunk 3 above)
Comment (ovisan) [inline chef/cookbooks/hyperv/recipes/openstack_install_nova.rb]: @jdsn i have reverted back to do the ruby_block, because it's using the already defined resource in chef search_file_replace_line.

Comment (tboerger) [inline chef/cookbooks/hyperv/recipes/openstack_install_nova.rb]: This won't work. Resource name attributes have to be unique.

**Comment by ovisan** [comment]
Depends on:
https://github.com/crowbar/crowbar-core/pull/101

File: chef/cookbooks/hyperv/attributes/default.rb [hunk 5]
```diff
@@ -53,24 +53,24 @@
 default[:openstack][:location] = 'C:\OpenStack'
 
 default[:openstack][:nova][:name] = "nova"
-default[:openstack][:nova][:version] = "2014.2.1"
-default[:openstack][:nova][:file] = "nova-2014.2.1.tar.gz"
+default[:openstack][:nova][:version] = "12.0.0.0rc2"
+default[:openstack][:nova][:file] = "nova-stable-liberty.tar.gz"
```
Comment (vuntz) [inline chef/cookbooks/hyperv/attributes/default.rb]: I'd feel more comfortable if we take a specific release (12.0.0.0) instead of the stable-liberty tarball. The reason is that we don't want to link to a random version from git.

We will still use the stable-liberty bits in the package; that's the "weird" bit where we "patch" the tarballs to use the version.

(same for the neutron & ceilometer tarballs)

File: chef/cookbooks/hyperv/attributes/default.rb [hunk 6]
```diff
@@ -80,10 +80,10 @@
...
-default[:service][:nova][:description] = "Service Wrapper for Openstack Nova Compute - Manages the Openstack Nova Compute as a Windows Service"
+default[:service][:nova][:description] = "Service Wrapper for Openstack Nova Compute"
 default[:service][:neutron][:name] = "neutron-hyperv-agent"
 default[:service][:neutron][:displayname] = "OpenStack Neutron Hyper-V Agent Service"
-default[:service][:neutron][:description] = "Service Wrapper for Openstack Neutron Hyper-V Agent - Manages the OpenStack Neutron Hyper-V Agent as a Windows Service"
-default[:service][:ceilometer][:name] = "ceilometer-compute-agent"
-default[:service][:ceilometer][:displayname] = "OpenStack Ceilometer Compute Agent Service"
-default[:service][:ceilometer][:description] = "Service Wrapper for Openstack Ceilometer Compute Agent - Manages the OpenStack Ceilometer Compute Agent as a Windows Service"
+default[:service][:neutron][:description] = "Service Wrapper for Openstack Neutron Hyper-V Agent"
+default[:service][:ceilometer][:name] = "ceilometer-agent-notification"
```
Comment (vuntz) [inline chef/cookbooks/hyperv/attributes/default.rb:92]: Why this change?

File: chef/cookbooks/hyperv/recipes/openstack_install_neutron.rb [hunk 7]
```diff
@@ -2,7 +2,7 @@
 
 cookbook_file "#{node[:cache_location]}#{node[:openstack][:neutron][:file]}" do
   source node[:openstack][:neutron][:file]
-  not_if { ::File.exists?(node[:openstack][:neutron][:installed]) }
+  not_if { ::File.exist?(node[:openstack][:neutron][:file]) }
```
Comment (vuntz) [inline chef/cookbooks/hyperv/recipes/openstack_install_neutron.rb]: Why?

File: chef/cookbooks/hyperv/recipes/openstack_install_nova.rb [hunk 8]
```diff
@@ -2,7 +2,7 @@
 
 cookbook_file "#{node[:cache_location]}#{node[:openstack][:nova][:file]}" do
   source node[:openstack][:nova][:file]
-  not_if { ::File.exists?(node[:openstack][:nova][:installed]) }
+  not_if { ::File.exist?(node[:openstack][:nova][:file]) }
```
Comment (vuntz) [inline chef/cookbooks/hyperv/recipes/openstack_install_nova.rb]: Same question.

File: chef/cookbooks/hyperv/templates/default/nova.conf.erb [hunk 9]
```diff
@@ -31,13 +28,21 @@ rabbit_virtual_host=<%= @rabbit_settings[:vhost] %>
...
-neutron_api_insecure=<%= @neutron_insecure ? 'True' : 'False' %>
+
+[neutron]
+url=<%= @neutron_protocol%>://<%= @neutron_server_host%>:<%= @neutron_server_port%>
+auth_strategy=keystone
+admin_tenant_name=<%= @keystone_settings['service_tenant'] %>
+admin_username=<%= @neutron_service_user%>
+admin_password=<%= @neutron_service_password%>
+admin_auth_url=<%= @keystone_settings['internal_auth_url'] %>
+api_insecure=<%= @neutron_insecure ? 'True' : 'False' %>
```
Comment (vuntz) [inline chef/cookbooks/hyperv/templates/default/nova.conf.erb]: Can you add something like: `region_name = <%= @keystone_settings['endpoint_region'] %>` ?

File: chef/cookbooks/hyperv/templates/default/nova.conf.erb [hunk 10]
```diff
@@ -31,13 +28,21 @@ rabbit_virtual_host=<%= @rabbit_settings[:vhost] %>
...
-neutron_auth_strategy=keystone
-neutron_admin_tenant_name=<%= @keystone_settings['service_tenant'] %>
-neutron_admin_username=<%= @neutron_service_user%>
-neutron_admin_password=<%= @neutron_service_password%>
-neutron_admin_auth_url=<%= @keystone_settings['internal_auth_url'] %>
-neutron_api_insecure=<%= @neutron_insecure ? 'True' : 'False' %>
+
+[neutron]
+url=<%= @neutron_protocol%>://<%= @neutron_server_host%>:<%= @neutron_server_port%>
+auth_strategy=keystone
```
Comment (vuntz) [inline chef/cookbooks/hyperv/templates/default/nova.conf.erb]: You can kill that line, it's deprecated now.

File: chef/cookbooks/hyperv/templates/default/nova.conf.erb [hunk 11]
```diff
@@ -7,10 +7,7 @@ cinder_catalog_info=volume:cinder:internalURL
 cinder_api_insecure=<%= @cinder_insecure ? 'True' : 'False' %>
 fake_network=true
 instances_path=<%= @instances_path%>
-<%= "glance_api_servers=#{@glance_server_protocol}://#{@glance_server_host}:#{@glance_server_port}" unless @glance_server_host.nil? %>
-<%= "glance_api_insecure=#{@glance_server_insecure ? 'True' : 'False'}" unless @glance_server_host.nil? %>
 use_cow_images=true
-force_config_drive=false
```
Comment (vuntz) [inline chef/cookbooks/hyperv/templates/default/nova.conf.erb:13]: Why remove this one?

File: chef/cookbooks/hyperv/attributes/default.rb (hunk 6 above)
Comment (ovisan) [inline chef/cookbooks/hyperv/attributes/default.rb:92]: because of the 100 characters limit hound is bitching about. Have a chat with @tboerger about that. Also the name is descriptive enough. Also the name for ceilometer agent changed, in fact they changed so much more than the name that it doesn't work the same as before and needs more work, that's why ceilometer is in another trello card.

File: chef/cookbooks/hyperv/recipes/openstack_install_neutron.rb (hunk 7 above)
Comment (ovisan) [inline chef/cookbooks/hyperv/recipes/openstack_install_neutron.rb]: the old way would re-install neutron with each chef run. Like this it won't

File: chef/cookbooks/hyperv/recipes/openstack_install_nova.rb (hunk 8 above)
Comment (ovisan) [inline chef/cookbooks/hyperv/recipes/openstack_install_nova.rb]: same answer

File: chef/cookbooks/hyperv/templates/default/nova.conf.erb (hunk 11 above)
Comment (ovisan) [inline chef/cookbooks/hyperv/templates/default/nova.conf.erb:13]: There was an error with it, can't recall which, but it's no surprise in openstack

File: chef/cookbooks/hyperv/attributes/default.rb (hunk 5 above)
Comment (ovisan) [inline chef/cookbooks/hyperv/attributes/default.rb]: hmm, there is no 12.0.0.0 in http://tarballs.openstack.org/neutron/ maybe neutron-2015.1.2.tar.gz ?
openstack way of naming their releases it's confusing.
It would help avoid unnecessary work if the preferences/requirements are specified ahead, preferably in the trello card as tasks.

File: chef/cookbooks/hyperv/templates/default/nova.conf.erb (hunk 9 above)
Comment (ovisan) [inline chef/cookbooks/hyperv/templates/default/nova.conf.erb]: Added.

File: chef/cookbooks/hyperv/templates/default/nova.conf.erb (hunk 10 above)
Comment (ovisan) [inline chef/cookbooks/hyperv/templates/default/nova.conf.erb]: Done.

File: .gitignore [hunk 12]
```diff
@@ -3,6 +3,7 @@
 
 .coverage/
 rubydeps.*
+*.swp
```
Comment (tboerger) [inline .gitignore]: :-1: This is something for your personal global gitignore. Currently all crowbar components share the same gitignore. Anyway this ignore is just related to your editor.

File: chef/cookbooks/hyperv/recipes/do_nova.rb [hunk 13]
```diff
@@ -2,6 +2,7 @@
 
 include_recipe "hyperv::openstack_install_neutron"
 include_recipe "hyperv::openstack_install_nova"
+include_recipe "hyperv::openstack_install_networking-hyperv"
```
Comment (tboerger) [inline chef/cookbooks/hyperv/recipes/do_nova.rb]: Please use an underscore

Comment (ovisan) [inline chef/cookbooks/hyperv/recipes/do_nova.rb]: i used a dash only because the package name has a dash http://tarballs.openstack.org/networking-hyperv/ but i will change it to underscore

File: chef/cookbooks/hyperv/recipes/openstack_install_neutron.rb (hunk 7 above)
Comment (dirkmueller) [inline chef/cookbooks/hyperv/recipes/openstack_install_neutron.rb]: but the point of the [:installed] attribute is to guard multiple runs. so the attribute needs fixing to a file that actually exists.

File: chef/cookbooks/hyperv/recipes/openstack_install_nova.rb (hunk 8 above)
Comment (dirkmueller) [inline chef/cookbooks/hyperv/recipes/openstack_install_nova.rb]: but the point of the [:installed] attribute is to guard multiple runs. so the attribute needs fixing to a file that actually exists.

**Comment by dirkmueller** [comment]
not before M8

File: README.md [hunk 14]
```diff
@@ -28,3 +28,14 @@ To get in contact with the developers you have multiple options, all of them are
 Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 specific language governing permissions and limitations under the License.
+
+
+## Note for manual intall/testing:
```
Comment (dirkmueller) [inline README.md]: install

File: chef/cookbooks/hyperv/recipes/do_nova.rb [hunk 15]
```diff
@@ -2,6 +2,7 @@
 
 include_recipe "hyperv::openstack_install_neutron"
 include_recipe "hyperv::openstack_install_nova"
+include_recipe "hyperv::openstack_install_networking_hyperv"
```
Comment (vuntz) [inline chef/cookbooks/hyperv/recipes/do_nova.rb:4]: Shouldn't that happen before (or just after) the neutron recipe?

**Comment by vuntz** [comment]
As mentioned on IRC a couple of days ago: I think we might want to install networking-hyperv (and maybe compute-hyperv, but it looks like nova still ships the code for that) in the same way we install nova, neutron and so on. But I'm not 100% sure and this should be double-checked.

File: chef/cookbooks/hyperv/recipes/openstack_install_networking_hyperv.rb [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+end
+
+powershell "install_networking_hyperv" do
+  code <<-EOH
+  cd #{node[:openstack][:location]}
+  cd #{node[:openstack][:networking_hyperv][:name]}
+  $env:PBR_VERSION=Get-Content setup.cfg | Select-String -Pattern "version = " | %{$_ -replace "version = ", ""}
+  #{node[:python][:command]} setup.py install
+  EOH
+  not_if { ::File.exist?("#{node[:openstack][:networking_hyperv][:installed]}") }
```
Comment (vuntz) [inline chef/cookbooks/hyperv/recipes/openstack_install_networking_hyperv.rb:25]: This attribute is not defined :/

**Comment by vuntz** [comment]
@ovisan I've done some fixes and cleanups at https://github.com/ovisan/crowbar-hyperv/pull/1. Unfortunately, I can't test things, so I'm not 100% sure it's fully correct. Can you check it out and merge if it's okay for you?

**Comment by vuntz** [comment]
More fixes in https://github.com/ovisan/crowbar-hyperv/pull/2 -- I was able to get the chef run to complete, and to get neutron and nova starting.

**Comment by vuntz** [comment]
We also depend on https://github.com/crowbar/crowbar-ha/pull/76

File: chef/cookbooks/hyperv/templates/default/nova.conf.erb [hunk 17]
```diff
@@ -28,19 +25,33 @@ rabbit_password=<%= @rabbit_settings[:password] %>
...
+instance_usage_audit = True
+instance_usage_audit_period = hour
+notify_on_state_change = vm_and_task_state
+
+[neutron]
+url=<%= @neutron_protocol %>://<%= @neutron_server_host %>:<%= @neutron_server_port %>
+admin_tenant_name=<%= @keystone_settings['service_tenant'] %>
+admin_username=<%= @neutron_service_user %>
+admin_password=<%= @neutron_service_password %>
+admin_auth_url=<%= @keystone_settings['internal_auth_url'] %>
```
Comment (dirkmueller) [inline chef/cookbooks/hyperv/templates/default/nova.conf.erb]: This should be

``` ruby
admin_auth_url = <%= KeystoneHelper.versioned_service_URL(@keystone_settings["protocol"],
                                      @keystone_settings["internal_url_host"],
                                      @keystone_settings["service_port"],
                                      "2.0") %>
```

Comment (ovisan) [inline chef/cookbooks/hyperv/templates/default/nova.conf.erb]: Thanks @dirkmueller

**Comment by ovisan** [comment]
latest-02-08-16-1454927372
latest-02-08-16-1454927408
screenshot-02-08-16-1454927570
Instance launched and it works, it gets an IP and it's running.

**Comment by vuntz** [comment]
@ovisan hrm, so you removed all the changes I had submitted? Was that expected?

**Comment by vuntz** [comment]
Anyway, +1 for this and I can submit a followup PR...

**Comment by ovisan** [comment]
@vuntz strange, i didn't meant to do that, sorry

**Comment by bmwiedemann** [comment]
+1
